libvcx/src/aries/handlers/connection/connection.rs

```diff
@@ -521,4 +521,381 @@ Get messages received from connection counterparty.
+        let _setup = SetupMocks::init();
+
+        let connection = Connection::create("test_serialize_deserialize", true);
+        let first_string = connection.to_string().unwrap();
+
+        let connection2 = Connection::from_string(&first_string).unwrap();
+        let second_string = connection2.to_string().unwrap();
+
+        assert_eq!(first_string, second_string);
+    }
+
+    pub fn create_connected_connections(consumer: &mut Alice, institution: &mut Faber) -> (Connection, Connection) {
+        debug!("Institution is going to create connection.");
+        institution.activate().unwrap();
+        let mut institution_to_consumer = Connection::create("consumer", true);
+        let _my_public_did = settings::get_config_value(settings::CONFIG_INSTITUTION_DID).unwrap();
+        institution_to_consumer.connect().unwrap();
+        let details = institution_to_consumer.get_invite_details().unwrap();
+
+        consumer.activate().unwrap();
+        debug!("Consumer is going to accept connection invitation.");
+        let mut consumer_to_institution = Connection::create_with_invite("institution", details.clone(), true).unwrap();
+
+        consumer_to_institution.connect().unwrap();
+        consumer_to_institution.update_state().unwrap();
+
+        debug!("Institution is going to process connection request.");
+        institution.activate().unwrap();
+        thread::sleep(Duration::from_millis(500));
+        institution_to_consumer.update_state().unwrap();
+        assert_eq!(VcxStateType::VcxStateRequestReceived as u32, institution_to_consumer.state());
+
+        debug!("Consumer is going to complete the connection protocol.");
+        consumer.activate().unwrap();
+        consumer_to_institution.update_state().unwrap();
+        assert_eq!(VcxStateType::VcxStateAccepted as u32, consumer_to_institution.state());
+
+        debug!("Institution is going to complete the connection protocol.");
+        institution.activate().unwrap();
+        thread::sleep(Duration::from_millis(500));
+        institution_to_consumer.update_state().unwrap();
+        assert_eq!(VcxStateType::VcxStateAccepted as u32, institution_to_consumer.state());
+
+        (consumer_to_institution, institution_to_consumer)
+    }
+
+    #[cfg(feature = "agency_pool_tests")]
+    #[test]
+    fn test_send_and_download_messages() {
+        let _setup = SetupLibraryAgencyV2::init();
+        let mut institution = Faber::setup();
+        let mut consumer = Alice::setup();
+
+        let (alice_to_faber, faber_to_alice) = create_connected_connections(&mut consumer, &mut institution);
+
+        institution.activate().unwrap();
+        faber_to_alice.send_generic_message("Hello Alice").unwrap();
+        faber_to_alice.send_generic_message("How are you Alice?").unwrap();
+
+        consumer.activate().unwrap();
+        alice_to_faber.send_generic_message("Hello Faber").unwrap();
+
+        // make sure messages has be delivered
```

Comment:
Personally, I would like to see either all tests commented, or none. But not a dealbreaker...